hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/common/AbstractCallAttempt.java

```diff
@@ -16,9 +16,233 @@
+import com.hedera.node.app.service.contract.impl.exec.gas.SystemContractGasCalculator;
+import com.hedera.node.app.service.contract.impl.exec.scope.HederaNativeOperations;
+import com.hedera.node.app.service.contract.impl.exec.scope.VerificationStrategies;
+import com.hedera.node.app.service.contract.impl.exec.scope.VerificationStrategy;
+import com.hedera.node.app.service.contract.impl.exec.systemcontracts.hts.AddressIdConverter;
+import com.hedera.node.app.service.contract.impl.hevm.HederaWorldUpdater;
+import com.swirlds.config.api.Configuration;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.BufferUnderflowException;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+
 /**
- * Base class for system contract call attempts.
- * @Future for now this abstract class is empty, subsequent PRs will add common call attempt functionality
- * once a HasCallAttempt class has been added.
+ * Base class for HTS and HAS system contract call attempts.
  */
-public class AbstractCallAttempt {}
+public class AbstractCallAttempt {
+    private final byte[] selector;
+    protected Bytes input;
+    private final Address authorizingAddress;
+    // The id of the sender in the EVM frame
+    protected final AccountID senderId;
+    private final Address senderAddress;
+    private final boolean onlyDelegatableContractKeysActive;
+    protected final HederaWorldUpdater.Enhancement enhancement;
+    private final Configuration configuration;
+    private final AddressIdConverter addressIdConverter;
+    private final VerificationStrategies verificationStrategies;
+    private final SystemContractGasCalculator gasCalculator;
+    private final List<CallTranslator> callTranslators;
+    private final boolean isStaticCall;
+
+    // If non-null, the address of a non-contract entity (e.g., account or token) whose
+    // "bytecode" redirects all calls to a system contract address, and was determined
+    // to be the redirecting entity for this call attempt
+    protected @Nullable final Address redirectAddress;
+
+    // too many parameters
+    @SuppressWarnings("java:S107")
+    public AbstractCallAttempt(
+            @NonNull final Bytes input,
+            @NonNull final Address senderAddress,
+            @NonNull final Address authorizingAddress,
+            final boolean onlyDelegatableContractKeysActive,
+            @NonNull final HederaWorldUpdater.Enhancement enhancement,
+            @NonNull final Configuration configuration,
+            @NonNull final AddressIdConverter addressIdConverter,
+            @NonNull final VerificationStrategies verificationStrategies,
+            @NonNull final SystemContractGasCalculator gasCalculator,
+            @NonNull final List<CallTranslator> callTranslators,
+            final boolean isStaticCall,
+            @NonNull Function redirectFunction) {
```

Comment:
`final`

Reply:
Also, can this `Function` be fully qualified here to show it comes from headlong and not from the JDK?